xarray/backends/zarr.py

```diff
@@ -312,40 +314,111 @@ def encode_variable(self, variable):
     def encode_attribute(self, a):
         return _encode_zarr_attr_value(a)
 
-    def prepare_variable(self, name, variable, check_encoding=False,
-                         unlimited_dims=None):
-
-        attrs = variable.attrs.copy()
-        dims = variable.dims
-        dtype = variable.dtype
-        shape = variable.shape
-
-        fill_value = attrs.pop('_FillValue', None)
-        if variable.encoding == {'_FillValue': None} and fill_value is None:
-            variable.encoding = {}
-
-        encoding = _extract_zarr_variable_encoding(
-            variable, raise_on_invalid=check_encoding)
-
-        encoded_attrs = OrderedDict()
-        # the magic for storing the hidden dimension data
-        encoded_attrs[_DIMENSION_KEY] = dims
-        for k, v in attrs.items():
-            encoded_attrs[k] = self.encode_attribute(v)
-
-        zarr_array = self.ds.create(name, shape=shape, dtype=dtype,
-                                    fill_value=fill_value, **encoding)
-        zarr_array.attrs.put(encoded_attrs)
-
-        return zarr_array, variable.data
+    def store(self, variables, attributes, check_encoding_set=frozenset(),
+              writer=None, unlimited_dims=None):
+        """
+        Top level method for putting data on this store, this method:
+          - encodes variables/attributes
+          - sets dimensions
+          - sets variables
+
+        Parameters
+        ----------
+        variables : dict-like
+            Dictionary of key/value (variable name / xr.Variable) pairs
+        attributes : dict-like
+            Dictionary of key/value (attribute name / attribute) pairs
+        check_encoding_set : list-like
+            List of variables that should be checked for invalid encoding
+            values
+        writer : ArrayWriter
+        unlimited_dims : list-like
+            List of dimension names that should be treated as unlimited
+            dimensions.
+            dimension on which the zarray will be appended
+            only needed in append mode
+        """
+
+        variables_to_encode = OrderedDict()
+        variables_encoded = OrderedDict()
+        ds = None
+        for vn, v in variables.items():
+            name = _encode_variable_name(vn)
+            if name in self.ds:
+                # existing variable, get its encoding
+                # and apply it to appended variable
+                variables_encoded[vn] = v
+                if ds is None:
+                    ds = open_zarr(self.ds.store, auto_chunk=False)
+                v.encoding = ds[vn].encoding
```

Comment:
I find the logic here a bit hard to follow. In general, the pattern of setting a variable to `None` and then assigning it on the first pass of a loop can lead to confusion (at least for me).

What about something like?

```python
existing_variables = set([vn for vn in variables if _encode_variable_name(vn) in self.ds])
if len(existing_variables) > 0:
    ds = open_zarr(self.ds.store, auto_chunk=False)
    variables_encoded = OrderedDict()
    for vn in existing_variables:
        ...
```